### PR TITLE
Add `ruff` in replacement of `isort` and `flake8`, and define YAML to check style

### PR DESCRIPTION
**Pull request by alvarobartt**
Hi @nateraw, good work with 🔥 `fuego`!

I read [your tweet](https://twitter.com/_nateraw/status/1621568997182382080) and, as you were looking for feedback, so I decided to submit this PR.

## What's in this PR?

This PR includes [`ruff`](https://github.com/charliermarsh/ruff) in replacement of both `flake8` and `isort` as it's way more efficient, see https://github.com/charliermarsh/ruff/blob/main/README.md.

Same has been done recently for 🤗`transformers` at https://github.com/huggingface/transformers/pull/21480 or 🤗`datasets` at https://github.com/huggingface/datasets/pull/5519

So main changes are listed below:
* ⚡️ Add `ruff` in replacement of `isort` and `flake8` (it's faster)

### CODE REVIEW

File: src/fuego/run_on_spaces.py
Comment (nateraw) [inline src/fuego/run_on_spaces.py:474]: Don't like this change, is there something we can do about it?

**Review by nateraw — COMMENTED** [review]
Happy to integrate `ruff` since we don't have a linter yet here. But I don't want the style to be different, if we can manage that. 

As for hatch, I'd prefer to just leave it as setuptools, as that's what I use in all my other repos so its what I'm most comfortable with. 

Sorry if these are annoying changes to make 😓 . Maybe something nice to add would be just:

- ruff
- a github workflow that runs style checks/linter checks. Preferably one that's mostly copied from one of the other big huggingface repos (`datasets`, `transformers`, `accelerate`, etc).

**Comment by nateraw** [comment]
btw thank you for the contribution!!! I really really appreciate you checking out the project ❤️

File: src/fuego/run_on_spaces.py
Comment (alvarobartt) [reply src/fuego/run_on_spaces.py:474]: Yeah sure, let me check PEP 👍🏻

**Comment by alvarobartt** [comment]
Ok makes sense @nateraw, then I'll rollback the changes to migrate to `hatch` and just try to keep `ruff` integration, plus a GitHub Workflow to check style 👍🏻

**Comment by alvarobartt** [comment]
Also another thing is that in my personal projects I tend to use `black --preview` instead of just `black`, as the `--preview` offers some other features such as the string-processing to split strings into multiple lines so as not to break the max line length, and some more. But as stated by them at https://black.readthedocs.io/en/stable/the_black_code_style/index.html#stability-policy, it's not considered stable, so we can probably skip it too!

**Comment by ofek** [comment]
Hey, maintainer of Hatch/Hatchling here! Let me know if you have any questions about switching, generally speaking configuring Hatchling is easier and less error-prone

File: src/fuego/run_on_spaces.py
Comment (ofek) [reply src/fuego/run_on_spaces.py:474]: This can certainly be disabled, but there are two benefits:

- Creation is faster
- String keys allow for characters disallowed as Python identifiers, like hyphens

**Comment by alvarobartt** [comment]
Hi @ofek! As @nateraw said, this project is still in an early development phase, so I assume most of the stuff he mentioned is because he's more comfortable with using `setuptools` and Python-dicts formatted way. Maybe at some point in a more mature phase of the project, replacing `setuptools` with `hatch`/`hatchling` could be considered. WDYT @nateraw

**Comment by nateraw** [comment]
sorry for delay here folks will get back to this next week, I hope. Will address all comments, etc. and can fix the merge conflicts if need be ❤️ 

Will read up on Hatch...having a maintainer in here for support definitely makes me happy to see and makes me lean more towards wanting to use it.

**Comment by alvarobartt** [comment]
Do you want me to open a separate PR for `hatch` and leave this one just for `ruff`? @nateraw

**Comment by nateraw** [comment]
That would be wonderful, @alvarobartt - sure.

**Comment by alvarobartt** [comment]
Hi @nateraw already done! I just kept the `ruff` part of the PR, I'll prepare a separate one for `hatch`, @ofek are you ok if I add a tag you in that PR so that you can support us too? Or are you too busy right now?

**Comment by ofek** [comment]
Feel free to tag me!

File: pyproject.toml
Comment (nateraw) [inline pyproject.toml:9]: Since `setup.cfg` was deleted, what happens to the unused ignore rules that were defined there? [E203](https://www.flake8rules.com/rules/E203.html) and [W503](https://www.flake8rules.com/rules/W503.html)

File: setup.py
Comment (nateraw) [reply setup.py:20]: will have more in future, I assume, with testing, etc.

File: Makefile
Comment (nateraw) [inline Makefile:10]: Let's revert this line back, if it doesn't break anything. See transformers [equivalent](https://github.com/huggingface/transformers/blob/4deaa534f54b4a1ce4fb6496af1f603229b157c6/Makefile#L68-L70).

Also will ruff work recursively here or are we missing `src` in addition to `.`?

**Review by nateraw — COMMENTED** [review]
couple notes

File: src/fuego/__init__.py
Comment (alvarobartt) [reply src/fuego/__init__.py:4]: I add it to avoid F401, since the imports on top are not used, you can either # noqa: F401 in the import line or just add `__all__`, whatever you prefer! :)

File: Makefile
Comment (alvarobartt) [reply Makefile:10]: Indeed the `--preview` flag is mainly for string formatting, but sure, that's not stable and should be removed. And yes, indeed `black` does the exploration recursively as `ruff`, so setting it to `.` means all the nested directories will be formatted/linted if Python files.

File: pyproject.toml
Comment (alvarobartt) [reply pyproject.toml:9]: True, I'll check this later, nice catch!

File: Makefile
Comment (nateraw) [reply Makefile:10]: awesome I didn't know that. thanks :)

File: src/fuego/__init__.py
Comment (nateraw) [reply src/fuego/__init__.py:4]: Let's mark with noqa for now...I'll issue a PR after yours with lazy loading that'll resolve the underlying problem :) (same as we do in `huggingface_hub`/`transformers`, etc)

Comment (nateraw) [reply src/fuego/__init__.py:4]: actually fine if you leave it like this for now too since I'll just change it afterwards anyways

File: pyproject.toml
Comment (nateraw) [reply pyproject.toml:9]: honestly not too picky, so if we want to fix this later I'm ok with it. If you look into it and its too time consuming I'm fine with moving to separate PR later to patch

Comment (nateraw) [reply pyproject.toml:9]: don't wanna block this PR much longer

File: Makefile
Comment (nateraw) [inline Makefile:10]:
```suggestion
	black .
```

File: pyproject.toml
Comment (alvarobartt) [reply pyproject.toml:9]: I'll have a look in a couple hours whenever I'm out of office, and then we can merge :) Thanks

**Comment by nateraw** [comment]
Go ahead and rerequest review when you feel you're done. I'll build/test locally then and merge.

**Review by nateraw — APPROVED** [review]
Made a quick change to mirror what was in transformers, now merging. Thanks again!!

**Pull request merged**
